<a href="https://colab.research.google.com/github/Likelipop/Data-mining/blob/main/Code1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <h2>PART 1: TUTORIAL </h2>

In [2]:
import pandas as pd
import numpy as np
from sklearn import preprocessing
import os

FILE_PATH = os.path.join(os.getcwd(),"data.csv")
df = pd.read_csv(FILE_PATH)
df.set_index("ID", inplace = True)
df


First Name Last Name  Age  Gender Department   Salary Date of joining
ID                                                                      
1        John       Doe  25.0      M      Sales  50000.0        1/1/2020
2        jane     Smith  30.0      F  Marketing  60000.0        6/1/2018
3         Bod   Johnson  45.0      M         HR  70000.0        9/1/2016
4       Alice  Williams  33.0      F         IT  80000.0        2/1/2017
5       James     Brown  27.0      M      Sales  55000.0        3/1/2019
6       Sarah       Lee   NaN      F  Marketing  65000.0       12/1/2018
7     Michael     Davis  39.0      M         HR      NaN        8/1/2015
8       Susan    Miller  42.0      F         IT   9000.0       11/1/2014
9       David    Wilson  28.0      M      Sales  60000.0        5/1/2020
10      Emily     Brown  35.0      F  Marketing  55000.0        4/1/2017
11       John       Doe  25.0      M      Sales  50000.0        1/1/2020
12       John       Doe  25.0      M      Sales  50000.0        1/1/2020

In [3]:
df.columns = [a.strip() for a in df.columns]

In [4]:
print(df.isnull().sum())

First Name         0
Last Name          0
Age                1
Gender             0
Department         0
Salary             1
Date of joining    0
dtype: int64


In [5]:

df["Age"] = df["Age"].fillna(df["Age"].mean())
df["Salary"] = df["Salary"].fillna(df['Salary'].mean())

In [6]:
print(df.duplicated().sum())

2


In [7]:
df = df.drop_duplicates()
print(df)

   First Name Last Name        Age Gender Department        Salary  \
ID                                                                   
1        John       Doe  25.000000      M      Sales  50000.000000   
2        jane     Smith  30.000000      F  Marketing  60000.000000   
3         Bod   Johnson  45.000000      M         HR  70000.000000   
4       Alice  Williams  33.000000      F         IT  80000.000000   
5       James     Brown  27.000000      M      Sales  55000.000000   
6       Sarah       Lee  32.181818      F  Marketing  65000.000000   
7     Michael     Davis  39.000000      M         HR  54909.090909   
8       Susan    Miller  42.000000      F         IT   9000.000000   
9       David    Wilson  28.000000      M      Sales  60000.000000   
10      Emily     Brown  35.000000      F  Marketing  55000.000000   

   Date of joining  
ID                  
1         1/1/2020  
2         6/1/2018  
3         9/1/2016  
4         2/1/2017  
5         3/1/2019  
6        12/

In [8]:
df = pd.get_dummies(df, columns = ['Gender','Department'], dtype ='int' )

In [9]:
df

First Name Last Name        Age        Salary Date of joining  Gender_F  \
ID                                                                           
1        John       Doe  25.000000  50000.000000        1/1/2020         0   
2        jane     Smith  30.000000  60000.000000        6/1/2018         1   
3         Bod   Johnson  45.000000  70000.000000        9/1/2016         0   
4       Alice  Williams  33.000000  80000.000000        2/1/2017         1   
5       James     Brown  27.000000  55000.000000        3/1/2019         0   
6       Sarah       Lee  32.181818  65000.000000       12/1/2018         1   
7     Michael     Davis  39.000000  54909.090909        8/1/2015         0   
8       Susan    Miller  42.000000   9000.000000       11/1/2014         1   
9       David    Wilson  28.000000  60000.000000        5/1/2020         0   
10      Emily     Brown  35.000000  55000.000000        4/1/2017         1   

    Gender_M  Department_HR  Department_IT  Department_Marketing  \
ID                                                                 
1          1              0              0                     0   
2          0              0              0                     1   
3          1              1              0                     0   
4          0              0              1                     0   
5          1              0              0                     0   
6          0              0              0                     1   
7          1              1              0                     0   
8          0              0              1                     0   
9          1              0              0                     0   
10         0              0              0                     1   

    Department_Sales  
ID                    
1                  1  
2                  0  
3                  0  
4                  0  
5                  1  
6                  0  
7                  0  
8                  0  
9                  1  
10                 0

In [10]:
df['Date of joining'] = pd.to_datetime(df['Date of joining'])
df['month'] = df['Date of joining'].dt.month
df['day_of_week'] = df['Date of joining'].dt.day_name()
df = df.drop('Date of joining', axis =1)
df

First Name Last Name        Age        Salary  Gender_F  Gender_M  \
ID                                                                     
1        John       Doe  25.000000  50000.000000         0         1   
2        jane     Smith  30.000000  60000.000000         1         0   
3         Bod   Johnson  45.000000  70000.000000         0         1   
4       Alice  Williams  33.000000  80000.000000         1         0   
5       James     Brown  27.000000  55000.000000         0         1   
6       Sarah       Lee  32.181818  65000.000000         1         0   
7     Michael     Davis  39.000000  54909.090909         0         1   
8       Susan    Miller  42.000000   9000.000000         1         0   
9       David    Wilson  28.000000  60000.000000         0         1   
10      Emily     Brown  35.000000  55000.000000         1         0   

    Department_HR  Department_IT  Department_Marketing  Department_Sales  \
ID                                                                         
1               0              0                     0                 1   
2               0              0                     1                 0   
3               1              0                     0                 0   
4               0              1                     0                 0   
5               0              0                     0                 1   
6               0              0                     1                 0   
7               1              0                     0                 0   
8               0              1                     0                 0   
9               0              0                     0                 1   
10              0              0                     1                 0   

    month day_of_week  
ID                     
1       1   Wednesday  
2       6      Friday  
3       9    Thursday  
4       2   Wednesday  
5       3      Friday  
6      12    Saturday  
7       8    Saturday  
8      11    Saturday  
9       5      Friday  
10      4    Saturday

In [11]:
df1 = df.drop(['First Name','Last Name','day_of_week'], axis= 1)
array = df1.values
print("Truoc khi bo cac gia tri nngoai lai\n",pd.DataFrame(array))

scaler = preprocessing.RobustScaler()
robust_df1 = scaler.fit_transform(array)
robust_df = pd.DataFrame(robust_df1)
robust_df
print("\n Sau khi bo cac gia tri ngoai lai\n", robust_df)

Truoc khi bo cac gia tri nngoai lai
            0             1    2    3    4    5    6    7     8
0  25.000000  50000.000000  0.0  1.0  0.0  0.0  0.0  1.0   1.0
1  30.000000  60000.000000  1.0  0.0  0.0  0.0  1.0  0.0   6.0
2  45.000000  70000.000000  0.0  1.0  1.0  0.0  0.0  0.0   9.0
3  33.000000  80000.000000  1.0  0.0  0.0  1.0  0.0  0.0   2.0
4  27.000000  55000.000000  0.0  1.0  0.0  0.0  0.0  1.0   3.0
5  32.181818  65000.000000  1.0  0.0  0.0  0.0  1.0  0.0  12.0
6  39.000000  54909.090909  0.0  1.0  1.0  0.0  0.0  0.0   8.0
7  42.000000   9000.000000  1.0  0.0  0.0  1.0  0.0  0.0  11.0
8  28.000000  60000.000000  0.0  1.0  0.0  0.0  0.0  1.0   5.0
9  35.000000  55000.000000  1.0  0.0  0.0  0.0  1.0  0.0   4.0

 Sau khi bo cac gia tri ngoai lai
           0         1    2    3    4    5         6         7         8
0 -0.799043 -0.850515 -0.5  0.5  0.0  0.0  0.000000  1.333333 -0.818182
1 -0.272727  0.283505  0.5 -0.5  0.0  0.0  1.333333  0.000000  0.090909
2  1.306220  1.417

In [ ]:
scaler = preprocessing.StandardScaler()
standard = scaler.fit_transform(array)
standard_df = pd.DataFrame(standard)
print("Chuan hoa du lieu \n", pd.DataFrame(standard))

scaler = preprocessing.MinMaxScaler()
minmax = scaler.fit_transform(array)
minmax_df = pd.DataFrame(minmax, index = df.index)
print('Scaling du lieu: \n', minmax_df)

Chuan hoa du lieu 
           0         1    2    3    4    5         6         7         8
0 -1.369782 -0.332811 -1.0  1.0 -0.5 -0.5 -0.654654  1.527525 -1.420508
1 -0.575077  0.232146  1.0 -1.0 -0.5 -0.5  1.527525 -0.654654 -0.027853
2  1.809037  0.797102 -1.0  1.0  2.0 -0.5 -0.654654 -0.654654  0.807740
3 -0.098254  1.362058  1.0 -1.0 -0.5  2.0 -0.654654 -0.654654 -1.141977
4 -1.051900 -0.050332 -1.0  1.0 -0.5 -0.5 -0.654654  1.527525 -0.863446
5 -0.228297  0.514624  1.0 -1.0 -0.5 -0.5  1.527525 -0.654654  1.643333
6  0.855391 -0.055468 -1.0  1.0  2.0 -0.5 -0.654654 -0.654654  0.529209
7  1.332214 -2.649132  1.0 -1.0 -0.5  2.0 -0.654654 -0.654654  1.364802
8 -0.892959  0.232146 -1.0  1.0 -0.5 -0.5 -0.654654  1.527525 -0.306384
9  0.219627 -0.050332  1.0 -1.0 -0.5 -0.5  1.527525 -0.654654 -0.584915
Scaling du lieu: 
            0         1    2    3    4    5    6    7         8
ID                                                            
1   0.000000  0.577465  0.0  1.0  0.0  0.0 

In [ ]:
df2 = standard_df.copy()
df2['equi_with_column0'] = pd.cut(x = df2[0], bins = 10)
print('Roi rac hoa cot 0 bang 10 equi width ranges: \n', df2)

Roi rac hoa cot 0 bang 10 equi width ranges: 
           0         1    2    3    4    5         6         7         8  equi_with_column0
0 -1.369782 -0.332811 -1.0  1.0 -0.5 -0.5 -0.654654  1.527525 -1.420508   (-1.373, -1.052]
1 -0.575077  0.232146  1.0 -1.0 -0.5 -0.5  1.527525 -0.654654 -0.027853   (-0.734, -0.416]
2  1.809037  0.797102 -1.0  1.0  2.0 -0.5 -0.654654 -0.654654  0.807740     (1.491, 1.809]
3 -0.098254  1.362058  1.0 -1.0 -0.5  2.0 -0.654654 -0.654654 -1.141977    (-0.0983, 0.22]
4 -1.051900 -0.050332 -1.0  1.0 -0.5 -0.5 -0.654654  1.527525 -0.863446   (-1.373, -1.052]
5 -0.228297  0.514624  1.0 -1.0 -0.5 -0.5  1.527525 -0.654654  1.643333  (-0.416, -0.0983]
6  0.855391 -0.055468 -1.0  1.0  2.0 -0.5 -0.654654 -0.654654  0.529209     (0.538, 0.855]
7  1.332214 -2.649132  1.0 -1.0 -0.5  2.0 -0.654654 -0.654654  1.364802     (1.173, 1.491]
8 -0.892959  0.232146 -1.0  1.0 -0.5 -0.5 -0.654654  1.527525 -0.306384   (-1.052, -0.734]
9  0.219627 -0.050332  1.0 -1.0 -0.5 -0.5  

In [ ]:
df3 = standard_df.copy()
df3['equi_depth_column0'] = pd.qcut(x = df3[0], q = 10)
print('Roi rac hoa cot 0 bang 10 equi width ranges: \n', df3)

Roi rac hoa cot 0 bang 10 equi width ranges: 
           0         1    2    3    4    5         6         7         8 equi_depth_column0
0 -1.369782 -0.332811 -1.0  1.0 -0.5 -0.5 -0.654654  1.527525 -1.420508   (-1.371, -1.084]
1 -0.575077  0.232146  1.0 -1.0 -0.5 -0.5  1.527525 -0.654654 -0.027853    (-0.67, -0.367]
2  1.809037  0.797102 -1.0  1.0  2.0 -0.5 -0.654654 -0.654654  0.807740      (1.38, 1.809]
3 -0.098254  1.362058  1.0 -1.0 -0.5  2.0 -0.654654 -0.654654 -1.141977   (-0.163, 0.0289]
4 -1.051900 -0.050332 -1.0  1.0 -0.5 -0.5 -0.654654  1.527525 -0.863446   (-1.084, -0.925]
5 -0.228297  0.514624  1.0 -1.0 -0.5 -0.5  1.527525 -0.654654  1.643333   (-0.367, -0.163]
6  0.855391 -0.055468 -1.0  1.0  2.0 -0.5 -0.654654 -0.654654  0.529209      (0.41, 0.951]
7  1.332214 -2.649132  1.0 -1.0 -0.5  2.0 -0.654654 -0.654654  1.364802      (0.951, 1.38]
8 -0.892959  0.232146 -1.0  1.0 -0.5 -0.5 -0.654654  1.527525 -0.306384    (-0.925, -0.67]
9  0.219627 -0.050332  1.0 -1.0 -0.5 -0.5  

# <h2>PART 2: BAI TAP VAN DUNG </h2>

# <h3>Load Library and Dataset</h3>

In [12]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import KBinsDiscretizer


In [13]:
arrhythmia_path = os.path.join(os.getcwd(), 'arrhythmia.zip')
!unzip {arrhythmia_path}


Archive:  /workspaces/Data-mining/Lab01/arrhythmia.zip
replace arrhythmia.data? [y]es, [n]o, [A]ll, [N]one, [r]ename: ^C


In [14]:
arhythmia_data_path = os.path.join(os.getcwd(), 'arrhythmia.data')
df = pd.read_csv(arhythmia_data_path, header = None, sep = ',')
df


0    1    2    3    4    5    6    7    8    9    ...  270   271   272  \
0     75    0  190   80   91  193  371  174  121  -16  ...  0.0   9.0  -0.9   
1     56    1  165   64   81  174  401  149   39   25  ...  0.0   8.5   0.0   
2     54    0  172   95  138  163  386  185  102   96  ...  0.0   9.5  -2.4   
3     55    0  175   94  100  202  380  179  143   28  ...  0.0  12.2  -2.2   
4     75    0  190   80   88  181  360  177  103  -16  ...  0.0  13.1  -3.6   
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   ...   ...   
447   53    1  160   70   80  199  382  154  117  -37  ...  0.0   4.3  -5.0   
448   37    0  190   85  100  137  361  201   73   86  ...  0.0  15.6  -1.6   
449   36    0  166   68  108  176  365  194  116  -85  ...  0.0  16.3 -28.6   
450   32    1  155   55   93  106  386  218   63   54  ... -0.4  12.0  -0.7   
451   78    1  160   70   79  127  364  138   78   28  ...  0.0  10.4  -1.8   

     273  274  275  276   277   278  279  
0    0.0  0.0  0.9  2.9  23.3  49.4    8  
1    0.0  0.0  0.2  2.1  20.4  38.8    6  
2    0.0  0.0  0.3  3.4  12.3  49.0   10  
3    0.0  0.0  0.4  2.6  34.6  61.6    1  
4    0.0  0.0 -0.1  3.9  25.4  62.8    7  
..   ...  ...  ...  ...   ...   ...  ...  
447  0.0  0.0  0.7  0.6  -4.4  -0.5    1  
448  0.0  0.0  0.4  2.4  38.0  62.4   10  
449  0.0  0.0  1.5  1.0 -44.2 -33.2    2  
450  0.0  0.0  0.5  2.4  25.0  46.6    1  
451  0.0  0.0  0.5  1.6  21.3  32.8    1  

[452 rows x 280 columns]

As observed above, the dataset has no column's name. We can refer to arrhythmia.names for additional information, which will help us initillize the column names later, then we should take a look at data's summary 

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 452 entries, 0 to 451
Columns: 280 entries, 0 to 279
dtypes: float64(120), int64(155), object(5)
memory usage: 988.9+ KB


There is 452 records, and each of them has 280 fields of data. Take a closer look, we may notice there is something wrong about dtypes : object(5). It gave me some intution about incorrect format data. Lets have a walk around at the data.

In [16]:
for column in df.columns:
    if df[column].dtype == 'object':
        print(df[column].head())

0    13
1    37
2    34
3    11
4    13
Name: 10, dtype: object
0     64
1    -17
2     70
3     -5
4     61
Name: 11, dtype: object
0    -2
1    31
2    66
3    20
4     3
Name: 12, dtype: object
0     ?
1     ?
2    23
3     ?
4     ?
Name: 13, dtype: object
0    63
1    53
2    75
3    71
4     ?
Name: 14, dtype: object


In [17]:
print("Giá trị ? :\n")
for column in df.columns:
    if (df[column] == '?').sum() :
       print(f"{column}:  {(df[column] == '?').sum()}")

Giá trị ? :

10:  8
11:  22
12:  1
13:  376
14:  1


So we replace "?" inside our dataset by np.nan for the ease of handling missing value.

In [18]:
df = df.replace('?', np.nan)

In [19]:
print("Giá trị ? :\n")
for column in df.columns:
    if (df[column] == '?').sum() :
       print(f"{column}:  {(df[column] == '?').sum()}")

Giá trị ? :



In [31]:
#check giá trị thiếu
print("Giá trị thiếu :\n",(df.isnull().sum() > 0).sum() )
#check giá trị trùng
print("Giá trị trùng: \n",df.duplicated().sum())


Giá trị thiếu :
 5
Giá trị trùng: 
 0


In [ ]:
# df = df.apply(pd.to_numeric, errors = 'coerce')

In [57]:
A= df.select_dtypes(include=['number'])
A.fillna(A.mean())

df[df.select_dtypes(include=['number']).columns] = A

In [56]:
j=2
for i in range(400):
    if pd.isnull(A.iloc[i,j]):
        print(f"i {i}  {type(A.iloc[i,j])}")

In [58]:
#check giá trị thiếu sau khi xử lý
for column in df.columns:
    if df[column].isnull().sum() > 0:
        print(f" columns {column}: {df[column].isnull().sum()}")

 columns 10: 8
 columns 11: 22
 columns 12: 1
 columns 13: 376
 columns 14: 1


In [69]:
A =df.iloc[:, 10:14]
A.value_counts()

10    11   12    13  
-115  54   11    -167    1
      63   -17   -148    1
-124  120  -135  -93     1
-136  75   63    -144    1
-144  28   61    -162    1
                        ..
84    39   70    114     1
86    29   69    103     1
      61   81    -117    1
87    30   -19   121     1
99    39   -29   144     1
Name: count, Length: 68, dtype: int64

Normalization and Scaling

In [58]:
array = df.values
# Outliers remove
scaler = preprocessing.RobustScaler()
robust_df = scaler.fit_transform(array)
robust_df= pd.DataFrame(robust_df)

#Normalization using z score
scaler = preprocessing.StandardScaler()
standard = scaler.fit_transform(array)
standard_df = pd.DataFrame(standard,index = df.index)
print("Chuan hoa du lieu: \n")
print(standard_df.head())


Chuan hoa du lieu: 

        0         1         2         3         4         5         6    \
0  1.734439 -1.107520  0.641327  0.713814  0.135505  0.844945  0.113709   
1  0.579312  0.902918 -0.031998 -0.251644 -0.516072  0.420769  1.013301   
2  0.457720 -1.107520  0.156533  1.618932  3.197915  0.175193  0.563505   
3  0.518516 -1.107520  0.237332  1.558590  0.721924  1.045871  0.383587   
4  1.734439 -1.107520  0.641327  0.713814 -0.059968  0.577044 -0.216141   

        7         8         9    ...       270       271       272       273  \
0  0.113809  1.201469 -1.094661  ...  0.508843 -0.013839  0.278621 -0.079546   
1 -0.588564 -1.977064 -0.191203  ...  0.508843 -0.157972  0.728573 -0.079546   
2  0.422853  0.464980  1.373324  ...  0.508843  0.130294 -0.471299 -0.079546   
3  0.254284  2.054247 -0.125096  ...  0.508843  0.908612 -0.371310 -0.079546   
4  0.198094  0.503742 -1.094661  ...  0.508843  1.168051 -1.071235 -0.079546   

   274       275       276       277       278 

In [59]:
#Max - Min Scaling
scaler = preprocessing.MinMaxScaler()
minmax = scaler.fit_transform(array)
minmax_df = pd.DataFrame(minmax, index = df.index)
print("\nMin max scalinng : \n")
print(minmax_df.head())


Min max scalinng : 

        0    1         2         3         4         5         6         7    \
0  0.903614  0.0  0.125926  0.435294  0.270677  0.368321  0.501805  0.241758   
1  0.674699  1.0  0.088889  0.341176  0.195489  0.332061  0.610108  0.150183   
2  0.650602  0.0  0.099259  0.523529  0.624060  0.311069  0.555957  0.282051   
3  0.662651  0.0  0.103704  0.517647  0.338346  0.385496  0.534296  0.260073   
4  0.903614  0.0  0.125926  0.435294  0.248120  0.345420  0.462094  0.252747   

        8         9    ...  270       271       272  273  274      275  \
0  0.590244  0.457478  ...  1.0  0.381356  0.968531  0.0  0.0  0.53125   
1  0.190244  0.577713  ...  1.0  0.360169  1.000000  0.0  0.0  0.31250   
2  0.497561  0.785924  ...  1.0  0.402542  0.916084  0.0  0.0  0.34375   
3  0.697561  0.586510  ...  1.0  0.516949  0.923077  0.0  0.0  0.37500   
4  0.502439  0.457478  ...  1.0  0.555085  0.874126  0.0  0.0  0.21875   

        276       277       278       279  
0  0.741

discretize 

In [ ]:
df2 = minmax_df.copy()
for column in df2.columns:
    df2[column + "equi_width"] = pd.cut(x = df2[column], bins= 10)


Roi rac hoa cot 0 bang 2 equi width ranges:

          0    1         2         3         4         5         6         7  \
0  0.903614  0.0  0.125926  0.435294  0.270677  0.368321  0.501805  0.241758   
1  0.674699  1.0  0.088889  0.341176  0.195489  0.332061  0.610108  0.150183   
2  0.650602  0.0  0.099259  0.523529  0.624060  0.311069  0.555957  0.282051   
3  0.662651  0.0  0.103704  0.517647  0.338346  0.385496  0.534296  0.260073   
4  0.903614  0.0  0.125926  0.435294  0.248120  0.345420  0.462094  0.252747   

          8         9  ...       271       272  273  274      275       276  \
0  0.590244  0.457478  ...  0.381356  0.968531  0.0  0.0  0.53125  0.741667   
1  0.190244  0.577713  ...  0.360169  1.000000  0.0  0.0  0.31250  0.675000   
2  0.497561  0.785924  ...  0.402542  0.916084  0.0  0.0  0.34375  0.783333   
3  0.697561  0.586510  ...  0.516949  0.923077  0.0  0.0  0.37500  0.716667   
4  0.502439  0.457478  ...  0.555085  0.874126  0.0  0.0  0.21875  0.825000   


In [67]:
df3 = minmax_df.copy()
for column in df3.columns:
    df3[str(column) + "equi_width"] = pd.qcut(x = df3[column], q= 10)


ValueError: Bin edges must be unique: Index([0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0], dtype='float64', name=1).
You can drop duplicate edges by setting the 'duplicates' kwarg